### PR TITLE
OpenAPI crawler

### PR DESCRIPTION
**Pull request by elic-eon**
### 🤔 Why?

To extract endpoints from OpenAPI spec.

### 🤓 What?

Add `openapi` crawler

### 🧪 Tested?

Test cases and verified end-to-end on dev.

### ☑️ Checks

- [x] My PR contains actual code changes, and I have updated the version number in `pyproject.toml`.

### CODE REVIEW

**Review by mars-lan — COMMENTED** [review]
LGTM at a high level. @usefulalgorithm @alyiwang pls take a closer look.

File: metaphor/openapi/extractor.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        if not openapi_json:
+            logger.error("Unable to get OAS json")
+            return []
+
+        endpoints = self._extract_paths(openapi_json)
+        hierarchies = self._build_hierarchies(openapi_json)
+
+        return hierarchies + endpoints
+
+    def _initial_session(self):
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:67]:
```suggestion
    def _initiailize_session(self):
```

File: metaphor/openapi/extractor.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        self._base_url = config.base_url
+        self._api_id = md5_digest(config.base_url.encode("utf-8"))
+        self._openapi_json_path = config.openapi_json_path
+        self._auth = config.auth
+        self._requests_session = requests.sessions.Session()
+
+    async def extract(self) -> Collection[ENTITY_TYPES]:
+        logger.info(f"Fetching metadata from {self._openapi_json_path}")
+
+        self._initial_session()
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:55]: move this to `__init__`

File: metaphor/openapi/extractor.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            return []
+
+        endpoints = self._extract_paths(openapi_json)
+        hierarchies = self._build_hierarchies(openapi_json)
+
+        return hierarchies + endpoints
+
+    def _initial_session(self):
+        if not self._auth:
+            return
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:69]: Is it OK to have no authentication for this crawler?

File: metaphor/openapi/extractor.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        if self._auth.basic_auth:
+            basic_auth = self._auth.basic_auth
+            self._requests_session.auth = (basic_auth.user, basic_auth.password)
+
+    def _get_openapi_json(self) -> Optional[dict]:
+        if not self._openapi_json_path.startswith("http"):
+            with open(self._openapi_json_path, "r") as f:
+                return json.load(f)
+
+        headers = OrderedDict(
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:80]: Do we still need this if all fields are empty?

File: metaphor/openapi/config.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+
+@dataclass(config=ConnectorConfig)
+class OpenAPIAuthConfig:
+    basic_auth: Optional[BasicAuth] = None
+
+
+@dataclass(config=ConnectorConfig)
+class OpenAPIRunConfig(BaseConfig):
+    openapi_json_path: str  # URL or file path
```
Comment (usefulalgorithm) [inline metaphor/openapi/config.py:22]: a path is a URL

File: metaphor/openapi/config.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+@dataclass(config=ConnectorConfig)
+class BasicAuth:
+    user: str
+    password: str
+
+
+@dataclass(config=ConnectorConfig)
+class OpenAPIAuthConfig:
+    basic_auth: Optional[BasicAuth] = None
```
Comment (usefulalgorithm) [inline metaphor/openapi/config.py:17]: are we going to have other kinds of authentication? If not then just put user and password here

File: metaphor/openapi/extractor.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    def _initial_session(self):
+        if not self._auth:
+            return
+
+        if self._auth.basic_auth:
+            basic_auth = self._auth.basic_auth
+            self._requests_session.auth = (basic_auth.user, basic_auth.password)
+
+    def _get_openapi_json(self) -> Optional[dict]:
+        if not self._openapi_json_path.startswith("http"):
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:76]: maybe use a url library like yarl to parse the url, and match against its scheme

File: metaphor/openapi/extractor.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+                "Connection": None,
+                "Accept-Encoding": None,
+            }
+        )
+        resp = self._requests_session.get(self._openapi_json_path, headers=headers)
+
+        if resp.status_code != 200:
+            return None
+
+        return resp.json()
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:93]: Is it possible to know the structure of this? If so let's do a pydantic model for it, if not then it's fine

File: metaphor/openapi/extractor.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+    def _extract_paths(self, openapi: dict) -> List[API]:
+        endpoints: List[API] = []
+        servers = openapi.get("servers")
+
+        for path, path_item in openapi["paths"].items():
+            path_servers = path_item.get("servers")
+            base_path = (
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:101]:
```suggestion
            server_url_path = (
```

joining `self._base_url` with `base_path` is a little confusing

File: metaphor/openapi/extractor.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                structure=AssetStructure(directories=[self._api_id], name=path),
+            )
+            endpoints.append(endpoint)
+        return endpoints
+
+    def _extract_methods(self, path_item: dict) -> List[OpenAPIMethod]:
+        def to_operation_type(method: str) -> Optional[OperationType]:
+            try:
+                operation_type = OperationType(method.upper())
+                return operation_type
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:126]:
```suggestion
                return OperationType(method.upper())
```

File: metaphor/openapi/config.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+
+@dataclass(config=ConnectorConfig)
+class OpenAPIRunConfig(BaseConfig):
+    openapi_json_path: str  # URL or file path
+    base_url: str  # base_url of endpoints
```
Comment (usefulalgorithm) [inline metaphor/openapi/config.py:23]: We can probably use Pydantic's URL types: https://docs.pydantic.dev/2.0/usage/types/urls/

File: metaphor/openapi/extractor.py (hunk 1 above)
Comment (elic-eon) [reply metaphor/openapi/extractor.py]: was intended to use `_init_*`

File: metaphor/openapi/extractor.py (hunk 4 above)
Comment (elic-eon) [reply metaphor/openapi/extractor.py:85]: Yes, `requests` will add these headers, we must set them to `None` explicitly

File: metaphor/openapi/config.py (hunk 5 above)
Comment (elic-eon) [reply metaphor/openapi/config.py]: Change the config for lesser confusion

File: metaphor/openapi/extractor.py (hunk 3 above)
Comment (elic-eon) [reply metaphor/openapi/extractor.py:73]: Yes, OAS could be hosted publicly

File: metaphor/openapi/config.py (hunk 6 above)
Comment (elic-eon) [reply metaphor/openapi/config.py:19]: Yes, we plan to have more authentication method

File: metaphor/openapi/extractor.py (hunk 8 above)
Comment (elic-eon) [reply metaphor/openapi/extractor.py:98]: Not going to model OpenAPI spec for now, might want to find existing package for OAS

File: metaphor/openapi/config.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+@dataclass(config=ConnectorConfig)
+class OpenAPIAuthConfig:
+    basic_auth: Optional[BasicAuth] = None
+
+
+@dataclass(config=ConnectorConfig)
+class OpenAPIRunConfig(BaseConfig):
+    base_url: HttpUrl
+    openapi_json_path: Optional[str] = None
```
Comment (usefulalgorithm) [inline metaphor/openapi/config.py:25]: https://docs.pydantic.dev/latest/api/types/#pydantic.types.FilePath

File: metaphor/openapi/extractor.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+            first_tag = self._get_first_tag(path_item)
+
+            endpoint = API(
+                logical_id=APILogicalID(
+                    name=endpoint_url, platform=APIPlatform.OPEN_API
+                ),
+                open_api=OpenAPI(path=path, methods=self._extract_methods(path_item)),
+                structure=AssetStructure(
+                    directories=[self._api_id] + [first_tag] if first_tag else [],
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:125]: why is the first tag a part of asset structure?

File: metaphor/openapi/extractor.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+            if not operation_type:
+                continue
+
+            methods.append(
+                OpenAPIMethod(
+                    summary=item.get("summary") or None,
+                    description=item.get("description") or None,
+                    type=operation_type,
+                )
+            )
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:158]:
```suggestion
            if operation_type := to_operation_type(method):
                methods.append(
                    OpenAPIMethod(
                        summary=item.get("summary"),
                        description=item.get("description"),
                        type=operation_type,
                    )
                )
```

File: metaphor/openapi/extractor.py [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+            f"Fetching metadata from {self._openapi_json_path or self._openapi_json_url}"
+        )
+
+        openapi_json = self._get_openapi_json()
+
+        if not openapi_json:
+            logger.error("Unable to get OAS json")
+            return []
+
+        endpoints = self._extract_paths(openapi_json)
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:64]:
```suggestion
        apis = self._extract_apis(openapi_json)
```

File: metaphor/openapi/extractor.py [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+        hierarchies = self._build_hierarchies(openapi_json)
```
Comment (usefulalgorithm) [inline metaphor/openapi/extractor.py:65]:
```suggestion
        hierarchies = self._extract_hierarchies(openapi_json)
```

**Review by usefulalgorithm — COMMENTED** [review]
almost there...

File: metaphor/openapi/extractor.py (hunk 13 above)
Comment (elic-eon) [reply metaphor/openapi/extractor.py:125]: Swagger uses a tag as a section but we don't multi-folder for entities, so we use the first tag as the folder. It works for most cases, a method usually only has one tag

**Review by usefulalgorithm — APPROVED** [review]
🚀

**Pull request merged**